File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int find(vector<int>& inorder, int target, int start, int end) {
        for(int i = start; i <= end; i++) {
            if(inorder[i] == target)
                return i;
        }
        return -1;
    }
 
    TreeNode* buildTreeHelper(vector<int>& preorder, vector<int>& inorder, 
                             int inStart, int inEnd, int& preIndex) {
     
        if(inStart > inEnd) {
            return NULL;
        }
        
        TreeNode* root = new TreeNode(preorder[preIndex]);
        int pos = find(inorder, preorder[preIndex], inStart, inEnd);
        preIndex++;
         
        root->left = buildTreeHelper(preorder, inorder, inStart, pos - 1, preIndex);
        root->right = buildTreeHelper(preorder, inorder, pos + 1, inEnd, preIndex);
        
        return root;
    }
 
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n = preorder.size();
        int preIndex = 0; 
        return buildTreeHelper(preorder, inorder, 0, n - 1, preIndex);
    }
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {

        return build(preorder, inorder,0 ,preorder.length - 1 ,0 ,inorder.length  -1);
        
    }
    public TreeNode build(int[] preorder, int[] inorder ,int preStart, int preEnd, int inStart, int inEnd){

        if(preStart > preEnd || inStart > inEnd){
            return null;
        } 

        int rootValue = preorder[preStart];

        TreeNode root = new TreeNode(rootValue);

        int index = inStart;

        while(inorder[index] != rootValue){
            index++;
        }
        int leftSize = index - inStart;

        root.left = build(preorder ,inorder ,preStart + 1 ,preStart + leftSize ,inStart, index - 1);

        root.right = build(preorder, inorder ,preStart + leftSize + 1 ,preEnd ,index + 1 ,inEnd);

        return root;
    }
}